### PR TITLE
feat: tag REVIEW.md prose and derive the review_md receipt (#289, #290, #291)

### LINKED ISSUE #289: phase2-triage.md 2l still says REVIEW.md can disable review dimensions; no such key exists
`references/phase2-triage.md` section 2l ("Determine Review Dimensions") opens with "All on by default unless REVIEW.md disables them." No REVIEW.md key disables a dimension: `review-md-spec.md` ("Rules and other prose") states that which dimensions run is decided by the trivial-scope gate alone and "There is no REVIEW.md key that changes this", and `phase1-preflight.md` (light-scope question) says the same. The 2l sentence is the last surviving pre-#94 claim (the Focus/Skip/per-dimension keys were retracted there).

Fix: rewrite the 2l opener to name the actual gate (trivial-scope light vs full, plus the two skip conditions already listed) and drop the REVIEW.md clause. Doc-only, suites-only. A doc-grep test could pin that no reference file claims REVIEW.md gates dimensions, if the phrase set is small enough to be a shape check rather than a phrase count.

Surfaced by the #196 recon (2026-09-04).

### LINKED ISSUE #290: Resolved-config receipt review_md=present|absent reflects the root REVIEW.md only; subdirectory-only configs render as absent
The resolved-config receipt line `review_md=present|absent (discovery)` (SKILL.md Phase 1 echo, `KNOB_REGISTRY` in `workflows/src/args.js`, rendered into the report's Review Methodology by `renderReport.js`) is sourced from the Phase 1 composite's `review_md_root` section, i.e. the repo-root REVIEW.md only. `validateArgs` enforces it against `reviewConfigPath` (root path). A repository whose only REVIEW.md files live in subdirectories therefore renders `review_md=absent` while those files govern thresholds and ignores for their subtrees (per-subtree scoping shipped in the #196 PR; before it, they governed the whole run and the receipt was equally wrong).

The machine-verifiable channel is right: the pipeline return carries `stats.reviewConfigSource`, `stats.reviewMdEntryCount` and (from #196) `stats.reviewMdSubtrees: [{dir, matched}]`. Only the human receipt is root-only.

Decision to make: (a) redefine the knob's value as "any REVIEW.md discovered" (source stays `discovery`; the skill stamps it after the 2d walk instead of the Phase 1 root check, and `validateArgs`'s `reviewConfigPath` cross-check changes), or (b) add a second registry row (e.g. `review_md_subtrees=<n>`) and keep the root row as is. Either touches `KNOB_REGISTRY`, `renderReport`, `render_report.test.js`, `test_boundary_parity.py` and the bench receipt parser, so it is its own PR. Suites-only.

Surfaced by the #196 red-team (2026-09-04); deliberately left out of that PR.

### LINKED ISSUE #291: REVIEW.md free prose is not subtree-scoped: every discovered file's rules reach every context-reading agent
`references/review-md-spec.md`'s inheritance model promises that `## Rules` and other free prose in a subdirectory REVIEW.md accumulate for files under that directory ("For a file in legacy/: rules = [rule-A, rule-B] (root only)"). The mechanism does not do that and never did: Phase 2d folds every discovered REVIEW.md's whole text, by value, into the one shared context file every discovery, validate and summarize agent reads, so `api/REVIEW.md`'s rules reach the agent reviewing `legacy/` exactly as root's do. The #196 PR scoped the parsed controls (thresholds, `ignore`) per subtree and rewrote the spec to say prose stays global advisory context, so the docs are now honest. This issue tracks whether prose should be scoped too.

Options: (1) ratify: prose is advisory and agents already weigh it against the file they are reading; (2) path-tag each REVIEW.md block in the context file (`<review-rules path="api/REVIEW.md">`, the shape `collect_project_rules.py` already uses for CLAUDE.md/AGENTS.md provenance in #82) and instruct agents to apply a block only to files under its directory: that is prompt-level scoping, so it is recall/noise-sensitive and rides a paired mini; (3) per-agent context slicing by changed-file directory, which conflicts with the single shared context file design.

Recommendation: (2) is the only structural option and is cheap on the writer side (the context file is already assembled by code); the instruction side is what needs measurement.

Surfaced by both #196 red-teams (2026-09-04).

### PR DESCRIPTION
**Pull request by leehopper**
## Why?

#289's 2l prose named a REVIEW.md dimension key that the implementation never provided, so the instructions described a control that could not exist. #290 kept the interactive receipt row derived from the root path alone, which rendered subdirectory-only discovery absent even though #196 made those configs govern their subtrees. #291 still folded REVIEW.md prose by value, leaving agents without a source path even though the spec, README, and wizard promised provenance.

## What Changed?

**Stage A**

- The `review_md` receipt row now derives from `DERIVED_FROM.reviewMd`, reporting `present` for any discovered root or subdirectory file and using the legacy `reviewConfigPath` fallback only when the raw array is absent.
- The `validateArgs` coherence rule now requires `reviewConfigPath` to be non-null exactly when an array entry has path `REVIEW.md`, while preserving subdirectory-only and legacy waist behavior.
- The generated contract fences in `scripts/resolve_config.py`, `skills/code-gauntlet/SKILL.md`, `skills/code-gauntlet/references/phase1-preflight.md`, and `skills/code-gauntlet/references/phase2-triage.md` were regenerated.
- Section 2l now names `deriveAgentFlags`, its light-scope behavior, the seven-agent default, and the model override policy, and a section-scoped contract pin covers the intended text.
- The Stage A fixture updates keep root `reviewConfigPath` coherent in delivery tests while leaving headless receipts without the interactive-only row.
- `tests/test_agent_contracts.py::TestReviewDimensionsContract::test_2l_names_the_dispatch_mechanism` pins the 2l mechanism and section boundary.
- `workflows/test/args.test.js` adds `validateArgs enforces reviewMd root and reviewConfigPath coherence`.
- `workflows/test/args.test.js` adds `T290: subdirectory discovery fills present and rejects absent`.
- `workflows/test/args.test.js` adds `T290: root discovery accepts present`.
- `workflows/test/args.test.js` adds `T290: empty discovery derives absent`.
- `workflows/test/args.test.js` adds `T290: legacy path preserves present`.
- `workflows/test/args.test.js` adds `T290: legacy null path refuses present`.
- `workflows/test/render_report.test.js` adds `T290-METH: subdirectory discovery renders the filled present receipt`.

**Stage B**

- `scripts/collect_project_rules.py` now discovers safe regular REVIEW.md files on the existing directory walk and renders accepted files as source-path-tagged `review-rules` blocks before `project-rules` blocks.
- REVIEW.md text is copied without parsing, import following, or cross-directory content deduplication, while caps and refusal paths remain disclosed through the collector receipt.
- Project imports that resolve to a rendered review source now record the `review_rules_source` skip and stop at that edge, while direct project walk hits, aliases, and out-of-walk Markdown imports retain their ruled behavior.
- The collector receipt now carries `review_md` metadata and `review_md_dirs` in deterministic walk order, and the caveat explains the advisory subtree scope of each review-rules block.
- The 2d step now reads the receipt's listed paths in receipt order for raw REVIEW.md waist inputs, derives the optional notice from receipt fields, and does not perform a second model-side walk.
- The context file's item (1) was removed from both the `skills/code-gauntlet/SKILL.md` and `skills/code-gauntlet/references/phase2-triage.md` twins so the collector output is folded from disk rather than REVIEW.md being retyped by value.
- The review-md spec, README, build-review-md wizard and practices, scaffold comments, and conventions-contract wording now describe source-tagged advisory prose and collector-owned discovery.
- `tests/test_collect_project_rules.py::TestReviewRules::test_root_and_subdirectory_blocks_precede_project_rules` pins review-rules ordering and tags.
- `tests/test_collect_project_rules.py::TestReviewRules::test_identical_review_files_remain_distinct_scopes` pins independent identical scopes and shared byte accounting.
- `tests/test_collect_project_rules.py::TestReviewRules::test_review_import_text_is_not_followed_or_skipped` pins that REVIEW.md imports remain literal prose.
- `tests/test_collect_project_rules.py::TestReviewRules::test_review_modified_flag_tracks_changed_paths` pins modified-in-this-diff metadata.
- `tests/test_collect_project_rules.py::TestReviewRules::test_review_receipt_lists_all_paths_without_text` pins receipt inventory, order, and text exclusion.
- `tests/test_collect_project_rules.py::TestReviewRules::test_review_only_repo_renders_caveat_and_block` pins review-only output and the unchanged project-rules absence gap.
- `tests/test_collect_project_rules.py::TestReviewRules::test_review_preserves_config_fence_and_normalizes_newlines` pins fence retention and newline handling.
- `tests/test_collect_project_rules.py::TestReviewRules::test_over_cap_review_is_discovered_but_never_opened` pins safe inventory without an over-cap read.
- `tests/test_collect_project_rules.py::TestReviewRules::test_total_cap_is_shared_and_discovery_survives_it` pins the shared review/project total cap.
- `tests/test_collect_project_rules.py::TestReviewRules::test_file_cap_bounds_review_reads_but_not_the_inventory` pins file-cap metadata and skips.
- `tests/test_collect_project_rules.py::TestReviewRules::test_review_symlink_keeps_scope_path` pins lexical scope paths and changed flags for symlinks.
- `tests/test_collect_project_rules.py::TestReviewRules::test_project_import_of_rendered_review_stops_at_that_target` pins the single import-membership guard and its gap behavior.
- `tests/test_collect_project_rules.py::TestReviewRules::test_review_alias_does_not_suppress_direct_project_source` pins that a review alias does not suppress a direct project source.
- `tests/test_collect_project_rules.py::TestReviewRules::test_out_of_walk_review_import_retains_project_import_behavior` pins out-of-walk REVIEW.md imports as ordinary project imports.
- `tests/test_collect_project_rules.py::TestReviewRules::test_review_receipt_directories_pin_walk_order_and_root` pins receipt directory order and root spelling.
- `tests/test_collect_project_rules.py::TestReviewRules::test_review_failure_receipt_retains_metadata_without_text` pins failure receipts and fallback output metadata.
- `tests/test_collect_project_rules.py::TestReviewRules::test_review_security_refusals_never_become_read_targets` pins lexical refusal paths and content-read refusal.
- `tests/test_collect_project_rules.py::TestFirstClassFileTypes::test_review_md_is_a_separate_source_kind` pins review-rules rendering and separate receipt metadata.
- `tests/test_collect_project_rules.py::TestDisclosureContract::test_repo_with_no_convention_files_succeeds_and_writes_the_empty_rules_notice` pins the no-source notice and the review-only exception.
- `tests/test_collect_project_rules.py::TestProvenance::test_caveat_is_first_and_emitted_once_only_when_sources_exist` pins conditional caveat wording.

## Additional Notes

Behaviour change: A subdirectory-only REVIEW.md now renders `review_md=present (discovery)`. The context file gains tagged `review-rules` blocks and loses the by-value item. The collector receipt gains `review_md` and `review_md_dirs`; legacy waists keep today's bytes. Bench receipts are headless, so the interactive row never appears there, while bench fixture repositories gain a root review-rules block in their context file. Filter behavior is unchanged and parity goldens are unchanged.

Rulings of record:

- **R2:** Keep one interactive `review_md` row and re-derive it from the authoritative `reviewMd` array, with a `reviewConfigPath` fallback only for a legacy waist. Option (b), a second receipt row, was rejected because `stats.reviewMdSubtrees` already carries subtree detail on the machine channel and the report does not act on a count.
- **R3:** Use option (2), collection in code inside the collector, and render review-rules blocks before project-rules blocks without parsing, following imports, or content deduplication. Option (1), ratifying prose as global, was rejected because the spec, README and wizard already promised provenance the by-value fold did not deliver; option (3), per-agent context slicing, conflicts with the single shared context file.
- **R4:** Run the collector before stamping raw REVIEW.md inputs and use its receipt for paths and order, with item (1) removed from both instruction twins. This gives discovery one authoritative source while preserving the raw text needed by the workflow waist.
- **A0-1:** Keep one realpath membership mechanism on the project import-follow edge only, and cover it with the alias and out-of-walk fixtures. This prevents a rendered review source from becoming a second project block while preserving direct project walk hits and legacy out-of-walk imports.
- **A0-4:** Keep `project_rules_absent` exactly as the project-source absence gap and use `review_md: []` as authoritative REVIEW.md absence. This preserves existing disclosure and bench scoring without adding a competing absence gap.
- **A0-6:** Add `review_md_dirs` in collector walk order with root spelled `.` and derive the optional notice from `review_md` plus `review_md_dirs`. This removes the need for a model-side directory walk while retaining the per-directory notice rule.
- **A0-7:** Enforce the iff rule between a root `REVIEW.md` entry and non-null `reviewConfigPath`. This keeps root provenance coherent while allowing subdirectory-only discovery and legacy callers.
- **A0-11:** Decline a per-entry digest until a consumer exists. The receipt should not grow a speculative field, so digest work remains an adjacent transcription-proof item.

Verification:

- `python -m pytest tests/ -q` -> `2264 passed, 2630 subtests passed in 75.76s (0:01:15)`.
- `python -m pytest bench/tests/ -q` -> `657 passed, 106 subtests passed in 33.63s`.
- `node --test workflows/test/*.test.js` -> `tests 1400; pass 1400; fail 0`.
- Python coverage -> `94.31%`, against the `93.8%` floor, with the required floor reached.
- Bench coverage -> `89.55%`, against the `88.5%` floor, with the required floor reached.
- JS coverage -> `99.10%` lines, `90.97%` branches, and `98.63%` functions, against floors `98.2%`, `90.0%`, and `97.7%`; the presence check exited 0.
- `node workflows/build.js && git diff --exit-code workflows/pipeline.js` -> `built 281622 bytes; pipeline-byte-identical-to-HEAD`.
- `python scripts/generate_contract_requirements.py --check` -> `generated registry blocks are current`.
- `python scripts/sync_agent_rules.py --check` -> `agent rule twins are current`.
- `pre-commit run --all-files` -> `All hooks passed, including ruff, format, mypy, markdownlint, cspell, and zizmor.`
- `claude plugin validate .` -> `Validation passed` with exit 0.

Mutation ledger:

| Mutation | Test that caught it | Result |
| --- | --- | --- |
| m1: restored the old REVIEW.md opener while retaining the mechanism sentence | `tests/test_agent_contracts.py::TestReviewDimensionsContract::test_2l_names_the_dispatch_mechanism` | Red, then restored green. |
| m2: removed `deriveAgentFlags` from section 2l | `tests/test_agent_contracts.py::TestReviewDimensionsContract::test_2l_names_the_dispatch_mechanism` | Red, then restored green. |
| m3: restored the old path-only fill | `T290: subdirectory discovery fills present and rejects absent`; `T290-METH: subdirectory discovery renders the filled present receipt` | Red, then restored green. |
| m4: removed the non-array legacy fallback | `T290: legacy path preserves present` | Red, then restored green. |
| m5: left the registry `derivedFrom` name as `reviewConfigPath` | `T5: derivation tables and registry names are bidirectionally complete` | Red, then restored green. |
| m15: deleted the reviewMd/reviewConfigPath iff coherence rule | `validateArgs enforces reviewMd root and reviewConfigPath coherence` | Red, then restored green. |
| Stage A array-arm variant: made every reviewMd array derive absent | `T290: root discovery accepts present` | Red, then restored green. |
| Stage A fallback variant: made every non-array case derive present | `T290: legacy null path refuses present` | Red, then restored green. |
| m6: reversed review-rules and project-rules render order | `tests/test_collect_project_rules.py::TestReviewRules::test_root_and_subdirectory_blocks_precede_project_rules` | Red, then restored green. |
| m7: renamed the review-rules tag to project-rules | `tests/test_collect_project_rules.py::TestReviewRules::test_root_and_subdirectory_blocks_precede_project_rules` | Red, then restored green. |
| m8: routed REVIEW.md through effective-content deduplication | `tests/test_collect_project_rules.py::TestReviewRules::test_identical_review_files_remain_distinct_scopes` | Red, then restored green. |
| m9: followed imports from `visit_review` | `tests/test_collect_project_rules.py::TestReviewRules::test_review_import_text_is_not_followed_or_skipped` | Red, then restored green. |
| m10: omitted `review_md` from the receipt | `tests/test_collect_project_rules.py::TestReviewRules::test_review_receipt_lists_all_paths_without_text` | Red, then restored green. |
| m11: rendered empty output when project sources were absent | `tests/test_collect_project_rules.py::TestReviewRules::test_review_only_repo_renders_caveat_and_block` | Red, then restored green. |
| m12: removed the conditional review caveat sentence | `tests/test_collect_project_rules.py::TestProvenance::test_caveat_is_first_and_emitted_once_only_when_sources_exist` | Red, then restored green. |
| m13: deleted the review-membership guard on project import-follow edges | `tests/test_collect_project_rules.py::TestReviewRules::test_project_import_of_rendered_review_stops_at_that_target` | Red, then restored green. |
| m14: suppressed direct visits whose realpath was a review source | `tests/test_collect_project_rules.py::TestReviewRules::test_review_alias_does_not_suppress_direct_project_source` | Red, then restored green. |
| m16: omitted `review_md_dirs` from the receipt | `tests/test_collect_project_rules.py::TestReviewRules::test_review_receipt_directories_pin_walk_order_and_root` | Red, then restored green. |
| m17: refused imports by lexical REVIEW.md basename | `tests/test_collect_project_rules.py::TestReviewRules::test_out_of_walk_review_import_retains_project_import_behavior` | Red, then restored green. |
| m18: used the resolved display path for `visit_review` | `tests/test_collect_project_rules.py::TestReviewRules::test_review_symlink_keeps_scope_path`; `tests/test_collect_project_rules.py::TestReviewRules::test_review_security_refusals_never_become_read_targets` | Red, then restored green. |
| m19: suppressed `project_rules_absent` for review-only repositories | `tests/test_collect_project_rules.py::TestDisclosureContract::test_repo_with_no_convention_files_succeeds_and_writes_the_empty_rules_notice` | Red, then restored green. |
| Fix 1 total-cap mutation: dropped the CLAUDE.md `total_cap_reached` skip | `tests/test_collect_project_rules.py::TestReviewRules::test_total_cap_is_shared_and_discovery_survives_it` | Red, then restored green. |
| Fix 1 m18 recheck: used the resolved display path for `visit_review` | `tests/test_collect_project_rules.py::TestReviewRules::test_review_security_refusals_never_become_read_targets` | Red, then restored green. |
| Fix 2: stripped YAML fence lines in `visit_review` | `tests/test_collect_project_rules.py::TestReviewRules::test_review_preserves_config_fence_and_normalizes_newlines` | Red, then restored green. |

Reviewer-designed mutation probes were also run. O1 made empty reviewMd derive present and was caught by `T182-ARGS: every registry receipt value rule and source rule is focused`, `validateArgs accepts an empty reviewMd array`, and `T290: empty discovery derives absent`; O2 kept only one direction of the coherence rule and was caught by `validateArgs enforces reviewMd root and reviewConfigPath coherence`; O3 sliced the 2l pin from heading 2k and was caught by `tests/test_agent_contracts.py::TestReviewDimensionsContract::test_2l_names_the_dispatch_mechanism`; O4 spelled the receipt root directory as an empty string and was caught by `tests/test_collect_project_rules.py::TestReviewRules::test_review_receipt_directories_pin_walk_order_and_root`; O5 inverted the review-source caveat condition and was caught by `tests/test_collect_project_rules.py::TestProvenance::test_caveat_is_first_and_emitted_once_only_when_sources_exist`; and O6 disclosed `review_rules_source` as an unresolved gap and was caught by `tests/test_collect_project_rules.py::TestReviewRules::test_project_import_of_rendered_review_stops_at_that_target`. The reviewer probe that charged each REVIEW.md twice against `total_bytes` was caught by `test_identical_review_files_remain_distinct_scopes` and `test_total_cap_is_shared_and_discovery_survives_it`; the resolved-path refusal probe was caught by `test_review_security_refusals_never_become_read_targets`; and the project-only caveat probe was caught by `TestProvenance::test_caveat_is_first_and_emitted_once_only_when_sources_exist`. One reviewer-designed probe survived green: x8a put text into the intermediate `review_md` entry, but the receipt strips text through `_sources_without_text`, so this was ruled non-defective; its x8b variant also returned that text in the receipt and went red under `test_review_receipt_lists_all_paths_without_text`. All other named mutations went red and were restored before the final gates.

Review shape: The reconnaissance and implementation plan were drafted by Codex Astra. One Claude Opus lens and one Codex Astra lens red-teamed the plan, both rejected the draft, and adjudication revised Stage B through A0-1 to A0-11. Two luna build stages produced the two-commit branch, `da8ebb4` for Stage A and `22ba433` for Stage B (amended with every fix round). The two-commit branch was reviewed by three Codex luna lenses and one Claude Opus lens: round 1 mutation returned accept, while conformance, docs, and Opus returned revise; round 2 conformance and docs returned revise on two docstring sentences and a test name; round 3 (docs) returned revise on two docstring sentences and the conventions agent's REVIEW.md checklist item; round 4 (docs, on the final head) returned revise on one minor only (a docstring sentence now also names replacement decoding and trailing-newline completion), which the orchestrator applied; no blocker or major remained. Fix 1 and Fix 2 ran on luna; the round-3 and round-4 edits were three sentences each, applied by the orchestrator and refuted by the next luna lens. This was a suites-only tier with no paired mini and no smoke; the bench mirror set contained zero REVIEW.md files.

Adjacent:

- To file: the raw-waist transcription proof for `reviewMd[].text`, with a digest only when a consumer exists.
- To file: a live Phase 2 smoke covering the collector receipt, receipt-order raw Reads, per-directory notice, and one collector-file fold.
- To file: the `review-md-spec.md` scope-gate wording at 87-93.
- To file: invalid-UTF-8 byte semantics for collector byte accounting.

- [x] Pipeline tests pass: `python -m pytest tests/ -q`
- [x] Bench self-tests pass: `python -m pytest bench/tests/ -q`
- [x] Workflow tests pass: `node --test workflows/test/*.test.js` (Node 24; the glob is required)
- [x] JS lint passes: `python3 workflows/test/tools/biome_check.py` (downloads the pinned Biome once)
- [x] Bundle rebuilt and byte-exact: `node workflows/build.js` then `git diff --exit-code workflows/pipeline.js`
- [x] Plugin manifests valid: `claude plugin validate .` (needs the Claude Code CLI; CI runs it regardless)
- [ ] Parity fixtures regenerated with `workflows/test/tools/record_parity.py` if a deterministic transform changed (not applicable: parity goldens unchanged)
- [x] Linters/hooks pass: `pre-commit run --all-files`
- [x] Docs, skill references, and agent contracts updated if behavior changed

Closes #289
Closes #290
Closes #291

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_0129xKEQRj3eEe2WXNrfQHyL

---

> [!NOTE]
> **Medium Risk**
> Changes Phase 2 context assembly and config receipt semantics agents rely on; behavior is heavily tested but mis-stamped args or collector failures now fail the handoff loudly.
> 
> **Overview**
> **REVIEW.md** is now a first-class, source-tagged context channel: `collect_project_rules.py` discovers safe `REVIEW.md` files on the existing walk, emits `<review-rules>` blocks (verbatim, no import following) before `<project-rules>`, and exposes `review_md` / `review_md_dirs` on the JSON receipt. Phase 2d and the shared context file use that collector as the single discovery authority—receipt-order raw reads for the workflow waist, unconditional fold from `--out`—instead of model-side globs or by-value REVIEW prose.
> 
> **Workflow contract fixes:** interactive `configEcho.review_md` derives from a nonempty `reviewMd` array (legacy `reviewConfigPath` only when the array is absent), so subdirectory-only configs report **present**; `validateArgs` requires root `REVIEW.md` iff `reviewConfigPath` is set. Docs, agents, README/spec, and 2l now describe subtree-advisory `review-rules`, collector-owned discovery, and `deriveAgentFlags` for dimension dispatch (not a fictional REVIEW.md dimension knob).
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 22ba4336c3631e4196b82a288b3a528981e3d564. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

**Comment by leehopper** [comment]
@liatrio-software-factory Approve

**Review by leehopper — COMMENTED** [review]
### ⚔️ Code Gauntlet

This PR (1189 changed lines) claims to restructure how REVIEW.md flows through the code-gauntlet review pipeline, closing a semantic gap where REVIEW.md prose reached agents as if globally shared rather than being attributed and scoped per subtree. Previously `scripts/collect_project_rules.py` deliberately excluded REVIEW.md from its collected "project rules" (it had a separate structured-parse path), and agent contracts described REVIEW.md's whole text as folded into shared context without per-directory scoping. The diff makes the collector walk for REVIEW.md alongside CLAUDE.md/AGENTS.md/QODO.md on the same directory pass, emitting each discovered REVIEW.md's raw text into a new `<review-rules path=... modified-in-this-diff=...>` block before `<project-rules>` blocks, and extends the collector's receipt with `review_md` (per-file metadata) and `review_md_dirs` (searched-directory walk order) fields plus a `review_rules_source` skip reason so an already-rendered REVIEW.md is not re-included as a project-rule import target. Companion changes update agent contracts (`bug-detector.md`, `code-simplifier.md`, `conventions-and-intent.md`), `README.md`, and multiple skill reference docs to adopt the new review-rules/project-rules vocabulary and have Phase 2d's discovery logic derive from the collector's receipt instead of an independent directory walk or Glob. On the workflow side, `workflows/src/args.js`/generated `workflows/pipeline.js` change the `review_md` knob's `derivedFrom` from `reviewConfigPath` to `reviewMd` and add a `validateArgs` coherence check requiring `reviewConfigPath` be non-null iff `args.reviewMd` contains an entry with path exactly `REVIEW.md`, with `scripts/resolve_config.py` updated to match; test files across Python and JS (including `workflows/test/stages_delivery.test.js`, which adds `reviewConfigPath: '/repo/REVIEW.md'` to four existing `runWith` tests as purely additive fixture data to satisfy the new coherence rule) are extended with coverage for the receipt fields, alias/symlink/security-refusal edge cases, and the new root/subdirectory coherence rule. Risk is concentrated in `scripts/collect_project_rules.py` and `scripts/resolve_config.py` (security-sensitive file-confinement/read logic) and in `workflows/pipeline.js`/`workflows/src/args.js` (the generated bundle and validation/derivation logic for a waist field consumed by every downstream review stage), all flagged high; the remaining touched files (README, agent prompts, skill references, and tests such as `stages_delivery.test.js`) are flagged low risk but are numerous and largely prose/doc/test changes whose accuracy depends on staying synchronized with the two high-risk implementation files.

2 finding(s) after the gauntlet — 2 low. 2 routed as improvement suggestion(s).

---
Generated by code-gauntlet | Reviewed up to: 168965510685ab899f5984ab9639f07a39d7d230

File: workflows/src/args.js [hunk 1]
```diff
@@ -1082,6 +1084,13 @@ export function validateArgs(args) {
           errors.push(`reviewMd[${i}].text must be a string`);
         }
       }
+      const hasRootReviewMd = args.reviewMd.some((entry) => (
+        entry && typeof entry === 'object' && !Array.isArray(entry) && entry.path === 'REVIEW.md'
+      ));
```
Comment (leehopper) [inline workflows/src/args.js:1092]: **💡 [LOW] New coherence check reinvents the existing isPlainObject helper**

The new hasRootReviewMd predicate manually re-spells the plain-object type guard (`entry && typeof entry === 'object' && !Array.isArray(entry)`) that the file already defines once as `isPlainObject` and reuses at over a dozen call sites. Before: `entry && typeof entry === 'object' && !Array.isArray(entry) && entry.path === 'REVIEW.md'`. After: `isPlainObject(entry) && entry.path === 'REVIEW.md'`. Using the shared helper keeps the object-shape definition in one place instead of two independently-maintained copies.

**Suggested fix:**
Call the existing helper instead of restating its logic: `const hasRootReviewMd = args.reviewMd.some((entry) => isPlainObject(entry) && entry.path === 'REVIEW.md');`

**Cited rule:**
> workflows/CLAUDE.md and AGENTS.md's 'Build the mechanism, not the instruction' / 'Extending should cost one edit' design principle — a shared shape guard already exists in this file and a new call site should use it rather than restate it.

```suggestion
      const hasRootReviewMd = args.reviewMd.some((entry) => (
        isPlainObject(entry) && entry.path === 'REVIEW.md'
      ));
```

⚔️ *Code Gauntlet*

File: workflows/src/args.js [hunk 2]
```diff
@@ -396,10 +396,12 @@ export const DERIVE_WHEN = {
...
-    fill: (args) => (args.reviewConfigPath != null ? 'present' : 'absent'),
+  reviewMd: {
+    fill: (args) => (Array.isArray(args.reviewMd)
+      ? (args.reviewMd.length > 0 ? 'present' : 'absent')
+      : (args.reviewConfigPath != null ? 'present' : 'absent')),
     source: 'discovery',
-    describe: '`present` when `reviewConfigPath` is set, else `absent`',
+    describe: '`present` when the `reviewMd` array is nonempty, else `absent`; without `reviewMd`, `present` when `reviewConfigPath` is set, else `absent`',
   },
 };
```
Comment (leehopper) [inline workflows/src/args.js:409]: **💡 [LOW] New DERIVED_FROM.reviewMd.fill uses a nested ternary**

The new fill function for the reviewMd derived-waist field nests one ternary inside each branch of an outer ternary to pick between four possible returns. Before: Array.isArray(args.reviewMd) ? (args.reviewMd.length > 0 ? 'present' : 'absent') : (args.reviewConfigPath != null ? 'present' : 'absent'). After: an if/else returning the same two-way ternary in each branch. This is the exact nested-ternary shape flagged as a readability trap; no other fill in this registry uses this style, so it is new complexity, not an established convention.

**Suggested fix:**
Replace the nested ternary with an if/else that returns early for the array case: `fill: (args) => { if (Array.isArray(args.reviewMd)) { return args.reviewMd.length > 0 ? 'present' : 'absent'; } return args.reviewConfigPath != null ? 'present' : 'absent'; }`. Remember this file is the source for the generated `workflows/pipeline.js` bundle (skip-listed as generated), so the identical fix belongs in the bundle only via `node workflows/build.js`, not a hand-edit.

```suggestion
export const DERIVED_FROM = {
  reviewMd: {
    fill: (args) => {
      if (Array.isArray(args.reviewMd)) {
        return args.reviewMd.length > 0 ? 'present' : 'absent';
      }
      return args.reviewConfigPath != null ? 'present' : 'absent';
    },
    source: 'discovery',
    describe: '`present` when the `reviewMd` array is nonempty, else `absent`; without `reviewMd`, `present` when `reviewConfigPath` is set, else `absent`',
  },
};
```

⚔️ *Code Gauntlet*

File: workflows/src/args.js (hunk 1 above)
Comment (leehopper) [reply workflows/src/args.js:1092]: Fixed in 4b1a75d: `hasRootReviewMd` now calls the existing `isPlainObject` helper.

File: workflows/src/args.js (hunk 2 above)
Comment (leehopper) [reply workflows/src/args.js:409]: Fixed in 4b1a75d: replaced the nested ternary with an if/else, exactly as suggested. Bundle rebuilt.
